src/Datasource/QueryTrait.php

```diff
@@ -552,6 +552,11 @@ public function __call(string $method, array $arguments)
     {
         $resultSetClass = $this->_decoratorClass();
         if (in_array($method, get_class_methods($resultSetClass), true)) {
+            deprecationWarning(sprintf(
+                'Unable to call `%s()` directly on query instance. ' .
```

Comment:
"Unable" sounds wrong since one is still able :)
An explicit message here might be quite annoying to users. Perhaps we should skip the message as we had done when deprecating `TableRegistry`'s method in 3.x.

Reply:
I was trying to say unable to on the query instance.

How would we deprecate it? Deprecate the __call() function?

Reply:
> How would we deprecate it? Deprecate the __call() function?

I just meant that the message needs to be tweaked.

Reply:
Suggestions on how to make it feel like TableRegistry?

Reply:
Hmm.. `TableRegistry` had concrete methods so `@deprecated` tag was enough. We could add the tag to the `__call()` method as you suggested. Not sure how IDE would deal with it but static analyzers should warn (if related rule/level is enabled).

We will also have to removed all the `@method` for methods of `CollectionInterface`.

Reply:
I think removing the `@method` annotations would make their static analysis fail though. Might be too harsh.

Reply:
I tried a couple things. Marking __call() as deprecated does not throw an error from psalm or my visual studio code extension. I'm not sure about phpstorm, but that's probably not good enough even if they do.

Reply:
> I think removing the @method annotations would make their static analysis fail though. Might be too harsh.

Having IDEs/tools report errors when using this magic method and not have any errors when actually running the code is our aim isn't?

Reply:
Is the concern with a deprecation warning that it will be too noisy? Thes method calls are going to be hard to track down without warnings or help from static analyzers.

Reply:
> Is the concern with a deprecation warning that it will be too noisy?

Yup

> Thes method calls are going to be hard to track down without warnings or help from static analyzers.

Yup, I guess keeping the explicit warning is the best option then.

Reply:
Would "Do not call..." sound better?

Reply:
```suggestion
                'Calling result set method `%s()` directly on query instance is deprecated. ' .
```

It should also explain why this method differs from others, i.e. it is not query, but result set/collection from here.

Reply:
The deprecation part is unnecessary in a deprecation warning I think.

Reply:
> The deprecation part is unnecessary in a deprecation warning I think.

We have had it in the various messages in the past and it read well. The message shown to users only has "Warning" prefix, so mentioning it's for deprecation is good.

Reply:
No worries. Any message you want is fine with me.